### ISSUE TITLE
Simulate offline mode

### ISSUE THREAD

**Issue by bahmutov** [cypress-io/cypress#235]
**Description**
Offline support is becoming important to web applications. It would be nice to switch the browser to the offline mode from Cypress either by test or for all tests and run it. If might be difficult and might require Cypress to handle ServiceWorkers very well.

**Additional Info**

**Comment by brian-mann** [comment]
This should be possible to do, but it's not immediately obvious or easy.

**Thoughts / Concerns**
- To date we haven't built in any specific API's that only work with one browser vendor. I'd like to try to find another way to do this, perhaps at OS network or proxy level since we control those two things.

- Chrome supports this out of the box in the `debugger` protocol but attempting to use that puts us in the same situation as issuing `native events` - that is you cannot have the `Dev Tools` open and an open connection to the `debugger` protocol. Chromium team is working on multiplexing support so we're kind of waiting for that to drop before adding support.

- If we did write API's exclusive to a single browser vendor, would they automatically throw (or be ignored) when running on a different browser? We'd also need a way at the `test` or `suite` level to indicate you want to *inclusively or exclusively* run tests based on the current browser.

- If we did use Chrome's APIs to achieve this, it would essentially break Cypress's ability to do automation since `websocket` events would no longer be able to reach the server. So if we did tap into Chrome we could only partially throttle and not fully force into offline mode. Another option would be to use the `WebRequest` API's and whitelist Cypress network traffic but throttle or reject others.

**Comment by jennifer-shehane** [comment]
With Chrome 63 coming out we will be able to support the debugger protocol. #832

**Comment by bbsimonbb** [comment]
Good onya guys. I'm ready for this right now. You know what to do.

**Comment by juristr** [comment]
Is there any progress/news on this? Would be cool to be able to test PWAs 👍

**Comment by jennifer-shehane** [comment]
No progress on this unfortunately.

**Comment by maku-zuhlke** [comment]
If you are using Redux, you can expose the store like this in eg index.js :
```
if (window.Cypress) {
    window.store = store
}
```
Then in a test in cypress you can set states/dispatch actions etc like:
```
cy.window().its('store')
            .then(
                store => store.dispatch({type: 'ACTION', status: false})
            );
```

**Comment by lolpez** [comment]
Guys it's 2019 👀

**Comment by mhrisse** [comment]
+1

**Comment by jennifer-shehane** [comment]
Related issue for clearing ServiceWorkers: https://github.com/cypress-io/cypress/issues/702

@lolpez We prioritize issues we work on based on a variety of factors - including the number of users requesting/helped by a feature. While this is still on our Roadmap, many other features are a higher priority today as a result of this assessment. Sorry for the delay.

**Comment by gr2m** [comment]
I tried to simulate offline by setting up a revers proxy to my running server and then shutting it down.

```js
const httpProxy = require("http-proxy");

context("Offline", () => {
  before(() => {
    httpProxy.createProxyServer({ target: "http://localhost:9000" }).listen(9001);

    cy.log("Open / on proxy");
    cy.visit("http://localhost:9001/");
  });

  it("should show update notification", () => {
    // wait for initial service worker installation
    cy.wait(3000);

    // stop the proxy server
    proxy.close();

    // reload the page
    cy.reload();
  });
});
```

But I get a `http.createServer is not a function` error. 

If I run just these two files in a separate file there is no problem

```js
const httpProxy = require("http-proxy");
httpProxy.createProxyServer({ target: "http://localhost:9000" }).listen(9001);
```

Is this a known limitation / problem?

**Comment by barnabynorman** [comment]
It would be really good to have this feature.

I can currently set the offline state in our vuex store but when changing pages (ie cy.visit) the application rechecks the state and updates the store value and is flushing queues before I get a chance to set the store again!

**Comment by miroslavvojtus** [comment]
Absence of this feature actually forces us to use selenium with direct access to driver api so we can test our core functionality in offline mode as our use case is offline-first.

I hope you will find time soon for this.

**Comment by Ozzyjtjustin** [comment]
The project my team is working on requires that we are able to simulate a No Internet Connection environment as well. 

Please provide an update on this!

**Comment by jpita** [comment]
hello, any updates on this?
is it on the roadmap?

**Comment by jennifer-shehane** [comment]
This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.

**Comment by shubhsherl** [comment]
One way of simulating offline mode:  
```
cy.server({ force404: true }); //  offline mode

cy.server({ enable: false});  // online mode
```
https://docs.cypress.io/api/commands/server.html

**Comment by jpita** [comment]
> One way of simulating offline mode:
> 
> ```
> cy.server({ force404: true }); //  offline mode
> 
> cy.server({ enable: false});  // online mode
> ```
> 
> https://docs.cypress.io/api/commands/server.html

that only blocks the calls, it doesn't tell the browser there's no internet

**Comment by shubhsherl** [comment]
> that only blocks the calls, it doesn't tell the browser there's no internet
Yes, it doesn't tell the browser that there is no internet, but by preventing the requests, we can test the offline functionality of PWA. It's just a workaround.

**Comment by jpita** [comment]
In my case it doesn't help, it should appear a "offline" banner and it doesn't. 
Thanks anyway

**Comment by shubhsherl** [comment]
> In my case it doesn't help, it should appear a "offline" banner and it doesn't.
> Thanks anyway

Have you tried reloading the site(`cy.reload()`) after switching to offline?

**Comment by EtienneBruines** [comment]
The root of the app is not cachable because of the MITM that Cypress performs (`document.domain = 'localhost'). 

This would make it (at least to me) that one would enable offline-mode *after* having started testing. First you'd want to install your service worker and such, adding all required files to the cache, and then run some tests. This would mean a global option is not required/desired - if you have no cache, no prior visits to the site - offline will never work anyways. 

## Design

```js
// Quick and dirty
cy.offline()
// - do tests inbetween
cy.online()

// A probably better way:
cy.network({offline: true})
// - do tests inbetween
cy.network({offline: false])
```

The latter would enable Cypress to at at later point also add throttling in the options. Not sure as to the usefulness of throttling, but that's point of discussion for another issue. 

## Docs

With the Chrome Debugger (Chrome, Electron), this would require to call:

```js
// possibly: await chromeDebugger.sendCommand(target, 'Network.enable');
await chromeDebugger.sendCommand(target, 'Network.emulateNetworkConditions', { offline: true });
```

## Implementation

I'm not really sure *where* this would be implemented in Cypress? 

There is some browser-specific implementations like `_maybeRecordVideo`, but not for all browsers (nor could I find a reliably call to these methods). https://github.com/cypress-io/cypress/blob/a22d0cc7563d7cb0cb176100cf93105557301e5a/packages/server/lib/browsers/electron.js#L81

Taking the screenshot-taking as an example. One would probably add a new command like https://github.com/cypress-io/cypress/blob/36866da64d86dc7e162489e357d7fc814ff025bb/packages/driver/src/cy/commands/screenshot.coffee

It has a dependency in its parent folder, https://github.com/cypress-io/cypress/blob/36866da64d86dc7e162489e357d7fc814ff025bb/packages/driver/src/cy/snapshots.coffee - I don't know why this is here? (why one would write code in one or the other?)

Then at some point there is the `automateScreenshot` method which calls `Cypress.automation("take:screenshot", props)` at some point. I take it this is where the command is being sent to the browser? 

There has been some progress into migrating code to vanilla js - I'm assuming this would be written in vanilla as well, and not in coffeescript?

@jennifer-shehane would you be able to give me some pointers as to which files/directories I could look at to implement this? (Meant to become a PR)

**Comment by lolpez** [comment]
Guys it's 2020 👀

**Comment by jpita** [comment]
> Guys it's 2020

Your point?

**Comment by rpocase** [comment]
@EtienneBruines Any chance this could take a staged approach with chromium based support first? There are already other examples in the docs of features/workarounds that are only available in chrome, and some coverage is better than none.

**Comment by EtienneBruines** [comment]
@rpocase if there is enough desire for it, I guess it could. 

As mentioned in #6932 we could publish it as a separate npm package (open source, so feel free, even if I don't end up doing so). The Cypress team has expressed (rightfully so) to not want this 'half-baked', so either full support in all by-Cypress-supported-browsers, or as a separate npm package that people can use at own risk. 

Today we submitted a PR to https://bugzilla.mozilla.org/show_bug.cgi?id=1553849 to enable `Network.emulateNetworkConditions` on Firefox as well. The people working there really were an awesome help in getting it this far. 

I can't speak on behalf of Cypress (just an outsider here :sweat_smile: ), but I'm guessing that once Firefox support lands (hopefully `Firefox 77.xxxx`, but I don't know for sure), and I add those tests to #6932 - they'd be up to including it. ( @jennifer-shehane ?)

**Comment by flotwig** [comment]
> Today we submitted a PR to [bugzilla.mozilla.org/show_bug.cgi?id=1553849](https://bugzilla.mozilla.org/show_bug.cgi?id=1553849) to enable `Network.emulateNetworkConditions` on Firefox as well. The people working there really were an awesome help in getting it this far.

@EtienneBruines awesome! we have been closely monitoring Firefox team's progress on the CDP implementation, we are excited to use it in Cypress

> I can't speak on behalf of Cypress (just an outsider here sweat_smile ), but I'm guessing that once Firefox support lands (hopefully `Firefox 77.xxxx`, but I don't know for sure), and I add those tests to #6932 - they'd be up to including it.

I *believe* CDP support is still only available in Firefox Nightly releases, so we can't really use *any* of it until they make it available in all editions of Firefox.

**Comment by EtienneBruines** [comment]
Quick update:

- `Network.emulateNetworkConditions` has landed in Firefox 78.0a1 nightly (probably tomorrow) :tada: 
- Firefox [always allows requests to `localhost`](https://bugzilla.mozilla.org/show_bug.cgi?id=1634246), as well as anything that resolves to it - even in offline mode (no patch in progress yet)
- I've managed to enable CDP for Firefox in Cypress, but this required two more changes:
    - `Browser.getVersion` update in Firefox (patch added, still being reviewed)
    - `Network.getAllCookies` support in Firefox (patch added, still being reviewed)

With all of the above patches in Firefox, I've managed to successfully implement `cy.network` on Firefox as well. Offline-only, so no throttling etc., but hey - that's all this issue is about. 

Regarding Chrome there's still the issue of service-workers ignoring offline emulation, but for the past two weeks others have been working hard on Chrome to fix [that issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1068544) as well - so I'm not worrying about that, that just means that service-worker support is only for newer Chrome versions.

Updated the PR with the progress - still lots of work to be done :sweat_smile:

**Comment by jpita** [comment]
@EtienneBruines is cy.network already available in the API?
if yes, can u point me to the docs?
haven't been able to find any info.

**Comment by EtienneBruines** [comment]
@jpita The underlying support for such a feature only just landed in Firefox Nightly, and Chrome is still working on some important bugfixes. 

That said, the Pull-Request (that is, #6932) was updated - but the PR has not yet been merged into `cypress/develop` and is therefore also not (yet) a part of any Cypress release. Sorry for the confusion.

The `cy.network` I mentioned is the one hypothesized in https://github.com/cypress-io/cypress/issues/235#issuecomment-608367340

**Comment by flotwig** [comment]
> Offline-only, so no throttling etc., but hey - that's all this issue is about.

FYI, #4176 will implement offline mode as well as simulated throttling/latency, but it's still a ways off. So having "offline mode" accessible via a plugin that uses CDP to get the job done will still be valuable in the mean time.

**Comment by EtienneBruines** [comment]
@flotwig That's nice to know. 

I'm just wondering, will #4176 affect the `window.onLine` representations, among others? Or is it a fancy way of turning the Cypress-Proxy off? (I don't know what the `window.onLine` value does when the proxy is offline.)

**Comment by flotwig** [comment]
@EtienneBruines From https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/onLine, it sounds like all browsers return `true` for `navigator.onLine` if the OS has any active network connection, regardless if they truly have access to the Internet or not. 

Since the browser will still be able to see the OS's network connection regardless of what the proxy layer is doing in #4176, I'm assuming #4176  will not affect `navigator.onLine`.

The HTML spec is pretty candid about this point: 

> This attribute [navigator.onLine] is inherently unreliable. A computer can be connected to a network without having Internet access.
> &mdash; https://html.spec.whatwg.org/multipage/offline.html#navigator.online

**Comment by EtienneBruines** [comment]
@AdrianoFerrari Which Cypress version is that? It seems like https://github.com/cypress-io/cypress/issues/8679, which should* be fixed in 5.6.0

*haven't verified this myself, but I trust the Cypress team.

**Comment by AdrianoFerrari** [comment]
I was using 5.3.0. Confirmed that with 5.6.0 I don't get the repeated/rapid retries of the image downloads. But the other failed requests (e.g. to `db/) also retry.

Closer, but I'm not sure how to proceed. Will post back with more details when I can.

**Comment by bahmutov** [comment]
We can do the offline mode during the test using Chrome debugger protocol, read https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode/

**Comment by EtienneBruines** [comment]
@bahmutov Do you know what the status is of exposing the Chrome debugger protocol for Firefox instances? The ball is in Cypress' court regarding that.

**Comment by bahmutov** [comment]
Firefox has certain limitations there, so that is still something we have not done

**Comment by lolpez** [comment]
Guys it's 2021 👀

**Comment by jennifer-shehane** [comment]
Reopening since this is not supported cross browser within the Cypress product. 

There is a workaround outlined here for Chromium browsers: https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode/

**Comment by Jacob-McKay** [comment]
Would really like to adopt the CDP solution, but it appears to only work when I target localhost in my test.  I verified this behavior with the cypress sample, am I crazy? https://github.com/cypress-io/cypress-example-recipes/issues/721

I'm going to try the fallback of mocking out window.navigator.onLine in the meantime

**Comment by jennifer-shehane** [comment]
You can use `cy.intercept('*', { forceNetworkError: true })` to also mimic offline mode in some circumstances, so that  is something  to explore.

**Comment by Kniaziev** [comment]
This one => https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode/ is not actual anymore, since last month, it started to fail after going offline, can't get back online, and all the next tests fails because it is still offline.

**Comment by SimonLandeholm** [comment]
The workaround now fails in chrome 96, when you wait intercepts with routeHandlers the request gets stuck in "pending" after going online again. I've tried cypress 8 and 9.

**Comment by lopezc184** [comment]
I was able to go offline and online again by using the code below. Had to remove the cypress-terminal-report plugin as well as I was getting an interesting and consistent logging error about ctrLogMessages.

```
CypressError
cy.task('ctrLogMessages') timed out after waiting 60000ms.

Because this error occurred during a after each hook we are skipping all of the remaining tests.Learn more
node_modules/cypress-terminal-report/src/collector/LogCollectSimpleControl.js:64:1
  62 |       cy.wait(wait, {log: false})
  63 |         .then(() => {
> 64 |           cy.task(
     | ^
  65 |             CONSTANTS.TASK_NAME,
  66 |             {
  67 |               spec: spec,
```
![image](https://user-images.githubusercontent.com/6199870/147331209-9dc5fa8a-dd5b-4db2-985f-f4476778169d.png)

```js
const assertOnline = () => {
  return cy.wrap(window).its('navigator.onLine').should('be.true')
}

const assertOffline = () => {
  return cy.wrap(window).its('navigator.onLine').should('be.false')
}

const goOffline = () => {
  return new Cypress.Promise(() => {
    return Cypress.automation('remote:debugger:protocol', {
      command: 'Network.enable'
    }).then(() => {
      return Cypress.automation('remote:debugger:protocol', {
        command: 'Network.emulateNetworkConditions',
        params: {
          offline: true,
          latency: -1,
          downloadThroughput: -1,
          uploadThroughput: -1
        }
      })
    })
  })
}

const goOnline = () => {
  return new Cypress.Promise(() => {
    return Cypress.automation('remote:debugger:protocol', {
      command: 'Network.emulateNetworkConditions',
      params: {
        offline: false,
        latency: -1,
        downloadThroughput: -1,
        uploadThroughput: -1
      }
    }).then(() => {
      return Cypress.automation('remote:debugger:protocol', {
        command: 'Network.disable'
      })
    })
  })
}

context('Network Test', () => {
  it('can go offline and back online', () => {
    // Check that we are online
    cy.visit('/')
    assertOnline()
    cy.get('some element').should('be.visible')
    
    // Go offline and verify
    cy.wrap(null).then(() => {
      goOffline()
      assertOffline()
      cy.get('some element').should('not.be.visible')
    })
    
    // Go back online and verify
    cy.wrap(null).then(() => {
      goOnline()
      assertOnline()
      cy.get('some element').should('be.visible')
    })
  })
})
```

**Comment by archfz** [comment]
@lopezc184 I was able to reproduce the issue, only on chrome and edge. This doesn't seem to be an issue with [cypress-terminal-report plugin](https://github.com/archfz/cypress-terminal-report). For example if you uninstall the plugin and add the following to your code, you will still reproduce the issue:

plugins.js
```js
module.exports = (on) => {
  on('task', {
    'mytask': () => console.log('lol'),
  });
};
```

end of your spec (the same code as you have above)
```js
   // ...
   cy.wrap(null).then(() => {
      goOnline()
      assertOnline()
      cy.get('some element').should('be.visible')
    })
   cy.task('mytask');
```

The issue seems to be with how cypress is communicating with the node js backend. Seemingly it uses network calls and messing with the network connection breaks it. That is what I think it is happening, but could be other cause as well.

**Comment by jhnns** [comment]
For documentation: The debugger protocol approach as described [on the blog](https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode/) doesn't work since 7.3.0. [This](https://github.com/cypress-io/cypress/issues/17723#issuecomment-906152925) is probably the cause.

**Comment by myasul** [comment]
I have implemented the following but nothing seems to work. Code can be seen below. 
- I used ` Cypress.automation('remote:debugger:protocol')`
- I also used `cy.intercept` with `req.destroy`

```typescript
export class NetworkConnectivity {
    private static readonly affectedUrlsRegex = /(?<=localhost:\d+)\/(apollo|api)\/.+$/g

    public static goOffline () {
        const offlineDevtoolCommand = {
            command: 'Network.emulateNetworkConditions',
            params: {
                offline: true,
                latency: -1,
                downloadThroughput: -1,
                uploadThroughput: -1
            }
        }

        return cy
            .then(() => Cypress.automation('remote:debugger:protocol', { command: 'Network.enable' }))
            .then(() => cy.task('sendCommandsToServiceWorker', { command: 'Network.enable' }))
            .then(() => cy.task('sendCommandsToServiceWorker', offlineDevtoolCommand, { timeout: 120000 }))
            .then(() => Cypress.automation('remote:debugger:protocol', offlineDevtoolCommand))
            .then(() => {
                // Reference: https://docs.cypress.io/api/commands/intercept#Controlling-the-response

                // @ts-ignore
                Cypress.config('network', { isOnline: false })

                cy.intercept(this.affectedUrlsRegex, req => {
                    // @ts-ignore
                    const { isOnline } = Cypress.config('network')

                    if (!isOnline) {
                        req.destroy()
                    }
                })
            })
    }
```

Both of these would not entirely intercept the request and the request would still return a 200 response. This might be because we are using service workers. To try and fix this issue, I tried using [`puppeteer-core`](https://github.com/puppeteer/puppeteer/issues/2469) to access the active service workers and send out the commands. This worked when tested locally but fails in CI/CD with the error `The Test Runner unexpectedly exited via a exit event with signal SIGSEGV` but without further explanations about the failure.

```javascript
const puppeteer = require('puppeteer-core')

const REMOTE_DEBUGGING_PORT = 9222

async function sendCommandsToServiceWorker ({ command, params }) {
    console.log(`Plugin started executing command ${command}, params: `, params)

    const browser = await puppeteer.connect({ browserURL: `http://localhost:${REMOTE_DEBUGGING_PORT}` })
    const targets = await browser.targets()
    const serviceWorkers = targets.filter((t) => t.type() === 'service_worker' || t.type() === 'other')

    const sendPromises = serviceWorkers.map(async (serviceWorker) => {
        const cdpSession = await serviceWorker.createCDPSession()
        await cdpSession.send(command, params)
    })

    await Promise.all(sendPromises)

    return null
}

module.exports = {
    sendCommandsToServiceWorker
}
```
Here is the screenshot of the CI/CD error.
![image](https://user-images.githubusercontent.com/25900564/151827454-d1b1a26c-edba-416a-8d66-bc98c975e80b.png)

I would really appreciate if anyone has solved this issue or have a workaround when using service workers.

**Comment by lolpez** [comment]
Guys it's 2022 👀

**Comment by aktibaba** [comment]
This worked for me 

cy.log('**go offline**'); 
    return Cypress.automation('remote:debugger:protocol',
    {
        command: 'Network.enable',
    }).then(() => {
        Cypress.automation('remote:debugger:protocol',
        {
            command: 'Network.emulateNetworkConditions',
            params: {
                offline: true,
                latency: -1,
                downloadThroughput: -1,
                uploadThroughput: -1,
            },
        })
        return cy.wait(1000);
    });

**Comment by laerteneto** [comment]
I **can** go **offline**, but I am **not** **able** to **get back online** again. So, following test cases fail.

However, it goes online/offline **without any problems** on **Electron**, but on **Chrome** and **Edge** it **does not**.

![image](https://user-images.githubusercontent.com/9399511/168798388-21db9b77-e8c9-4e36-aecc-a455a13ae85d.png)

**Comment by aktibaba** [comment]
Copy exactly same from above and try again

**Comment by laerteneto** [comment]
I tried, but it still does not work on Chrome when I am trying to get back online, it only works on Electron.

I also tried the solution posted in here, but without success:
https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode/

> Copy exactly same from above and try again

**Comment by aktibaba** [comment]
your code is missing return, premise returning nothing

**Comment by laerteneto** [comment]
> your code is missing return, premise returning nothing, try this
> 
> cy.log('go offline'); return Cypress.automation('remote:debugger:protocol', { command: 'Network.enable', }).then(() => { Cypress.automation('remote:debugger:protocol', { command: 'Network.emulateNetworkConditions', params: { offline: true, latency: -1, downloadThroughput: -1, uploadThroughput: -1, }, }) return cy.wait(1000); });

I managed to get it offline, but the problem is to put it back online, in which it is working well only on Electron.

**Comment by ifn** [comment]
Downgraded Cypress to 7.2.0. Seems to work.

**Comment by papamishka911** [comment]
Same problem: 7.9.0 version.

**Comment by DariusNorv** [comment]
I observe this problem in 10+ version in Electron runner as well.

**Comment by maximkoshelenko** [comment]
This one works for me

const goOffline = () => {
  cy.log('go offline');
  return Cypress.automation('remote:debugger:protocol',
    {
      command: 'Network.enable',
    }).then(() => {
      Cypress.automation('remote:debugger:protocol',
        {
          command: 'Network.emulateNetworkConditions',
          params: {
            offline: true,
            latency: -1,
            downloadThroughput: -1,
            uploadThroughput: -1,
          },
        })
      return true
    });
};

**Comment by Shnrqpdr** [comment]
I'm facing this problem too. I'm using this code below:

``` js
network(params) {
    cy.log(`**Offline: ${params.offline}**`)
      .then(() => {
        Cypress.automation('remote:debugger:protocol', {
          command: 'Network.enable',
        });
      })
      .then(() => {
        Cypress.automation('remote:debugger:protocol', {
          command: 'Network.emulateNetworkConditions',
          params: {
            offline: params.offline,
            latency: 0,
            downloadThroughput: 0,
            uploadThroughput: 0,
            connectionType: 'none',
          },
        });
      });
  },

```

I've work with cypress 9.5.1. After going offline, can't get back online. Anyone has a solution for this?

**Comment by asmirnov-style** [comment]
Cypress v10.3.1 and still the same issue with going back Online. 
Looks like it's turn Network for a second and then break connection.
Any ideas?

**Comment by Shnrqpdr** [comment]
Hello everyone

I have been stuck into this issue for at least 2 weeks ago and i got a "solution" to try simulate offline mode. The code is in the image below

![image](https://user-images.githubusercontent.com/86846589/214145859-d19a95d5-e657-41e4-b4eb-8d8b8460aae7.png)

Create this function on custom commands file of your project and call it using "cy.network({online: false})"
This work for me.
I hope that this workaround work for you too.

I'm using cypress 9.5.1

**Comment by lolpez** [comment]
Guys it's 2023 👀

**Comment by mostafijur022** [comment]
[laerteneto](https://github.com/laerteneto)
@laerteneto did you get any solution?

**Comment by buffcode** [comment]
Modifying the real `onLine` attribute, tested in Cypress 12.11.0:

```js
// go offline
cy.window().then((win) => {
  cy.stub(win.navigator, 'onLine').value(false);
  cy.wrap(win).trigger('offline');
});

// go online
cy.window().then((win) => {
  cy.stub(win.navigator, 'onLine').value(true);
  cy.wrap(win).trigger('online');
});
```

**Comment by antolinaproducepay** [comment]
@laerteneto did you get any solution? I m stuck with the same

**Comment by AntolinaGonzalez** [comment]
@laerteneto did you get any solution? Im stuck with go online too

**Comment by maxim-koshelenko** [comment]
> Modifying the real `onLine` attribute, tested in Cypress 12.11.0:
> 
> ```js
> // go offline
> cy.window().then((win) => {
>   cy.stub(win.navigator, 'onLine').value(false);
>   cy.wrap(win).trigger('offline');
> });
> 
> // go online
> cy.window().then((win) => {
>   cy.stub(win.navigator, 'onLine').value(true);
>   cy.wrap(win).trigger('online');
> });
> ```

Confirming the workability of this solution.

Cypress v10, Chrome 117 (only in headless).

Don't forget to put go online into before, beforeEach and afterEach

**Comment by lolpez** [comment]
Guys it's 2024 👀

**Comment by svetlana-napris-elucid** [comment]
I'm using this for Cypress 13.11.0 and Chrome v126: https://www.cypress.io/blog/2020/11/12/testing-application-in-offline-network-mode 
Works for me.

**Comment by gabscogna** [comment]
8 years later....

I did exactly what it's in this article, but the problem is that all tests after `goOnline` become absolutely slowly and I can't figure why.
It seems like `Network.disable` doesn't work.

**Comment by lolpez** [comment]
Guys it's 2025 👀

**Comment by Shnrqpdr** [comment]
Guys it's 2026

**Comment by lolpez** [comment]
Guys it's 2026 👀
